### PR TITLE
CLS Silverstone: enable the init sequence for all of the QSFPDD CMIS4…

### PR DESCRIPTION
**Pull request by ds952811**
… modules

Signed-off-by: Dante Su <dante.su@broadcom.com>

**- What I did**
1. Enable the initialization sequence for all of the QSFPDD CMIS4 modules, which used to be limited to InnoLight only.
2. Stop issuing software reset to the QSFPDD DACs even though its module state is not ModuleReady for 3+ seconds.
3. Lower the clock rate of FPGA-I2C to have the best transceiver compatibility.

**- How I did it**
See What I did

**- How to verify it**
1. Have both InnoLight and Eoptolink CMIS4 modules attached to the DUT
2. Issue 'show interfaces transceiver eeprom' to verify if their application selection is '2'
3. Intentionally disable anyone of the 4 links of the transceiver, and make sure the other 3 links are still UP

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by ds952811** [comment]
**PLEASE DO NOT** approve this review until the Broadcom QA finished the internal tests, I'll later send an all-clear flag once it's done.

This PR is early access to review what are the proposed changes to address the CMIS4 issues, while the new test image has yet generated, the DUT with corresponding changes manually applied are verified, please refer to the attached log for details.
[PR95-ENG-TEST.log](https://github.com/zhenggen-xu/sonic-buildimage/files/5594432/PR95-ENG-TEST.log)

File: platform/broadcom/sonic-platform-modules-cel/silverstone/modules/switchboard.c
Comment (ds952811) [inline platform/broadcom/sonic-platform-modules-cel/silverstone/modules/switchboard.c]: Honestly, this is unlikely to be the root cause of I2C failure reported yesterday, however it's the only thing under our control without having the Celestica engaged, that's why I did this.
Please note it's always working on our test unit no matter if high-speed (BIT6-0x40) is enabled or not.

File: device/celestica/x86_64-cel_silverstone-r0/plugins/sfputil.py
Comment (ds952811) [inline device/celestica/x86_64-cel_silverstone-r0/plugins/sfputil.py]: This is to address the DAC failure below:
pmon#sfputil.py: write failed: [Errno 110] Connection timed out

Some of the DAC (Passive Copper) do not support software reset, and hence such a write attempt will fail

File: device/celestica/x86_64-cel_silverstone-r0/plugins/sfputil.py
Comment (ds952811) [inline device/celestica/x86_64-cel_silverstone-r0/plugins/sfputil.py:531]: All of the CMIS3 modules we have are working fine with 4x100G in its default application mode, and hence this init sequence will only be performed on CMIS4 modules.

Note: The definition of the datapath initialiation slightly differs from CMIS3 to CMIS4, and hence we can't use _init_cmis4_module() on CMIS3 modules.

File: device/celestica/x86_64-cel_silverstone-r0/plugins/sfputil.py
Comment (ds952811) [inline device/celestica/x86_64-cel_silverstone-r0/plugins/sfputil.py]: The UT result shows that we need a few seconds for the actual configuration status correctly reflected on the register, but we don't really want to wait too long in the xcvrd, so I slightly enlarge the delay from 500ms to 1sec, hope this could reduce the redudant errors in the syslog

File: device/celestica/x86_64-cel_silverstone-r0/plugins/sfputil.py
Comment (ds952811) [inline device/celestica/x86_64-cel_silverstone-r0/plugins/sfputil.py:439]: There is no guarantee that 100G PAM4 mode is always the application 2, and hence we'll need to decode the application advertisement and figure the correct application code

File: device/celestica/x86_64-cel_silverstone-r0/plugins/sfputil.py
Comment (zhenggen-xu) [reply device/celestica/x86_64-cel_silverstone-r0/plugins/sfputil.py]: In case of DAC, should we even add failure counter and come here,  See line 553:
`self.mod_failure[x] += 1`

File: platform/broadcom/sonic-platform-modules-cel/silverstone/modules/switchboard.c
Comment (zhenggen-xu) [reply platform/broadcom/sonic-platform-modules-cel/silverstone/modules/switchboard.c]: If we turn off the high speed mode,  how slow would it be?  I don't feel we should do this unless we feel it was necessary.

File: device/celestica/x86_64-cel_silverstone-r0/plugins/sfputil.py
Comment (zhenggen-xu) [reply device/celestica/x86_64-cel_silverstone-r0/plugins/sfputil.py]: another generic question,  it seems we are keep monitoring all the settings and do reset/init etc all the time.  It is a bit concerning if we that could be disruptive.  Thinking maybe we could only do these steps when we got the optical insert event,  and finish the bring up process once.  of course, we will do it again if insert happens again. Let me know.

File: platform/broadcom/sonic-platform-modules-cel/silverstone/modules/switchboard.c
Comment (ds952811) [reply platform/broadcom/sonic-platform-modules-cel/silverstone/modules/switchboard.c]: 1. We will have ~30% delays for each I2C operations when the BIT6 (high-speed) is cleared
2. Based on the latest test report from Broadcom QA, slowing down the clock does not improve the transceiver compatibility.
And hence, I'll revert this change.

Note: None of the QSFPDD optics is available at my side, so I'm counting on QA for the verifications

e.g.
Dante,

It appears Eoptolink 3.0 broke. It took ~17 minutes for link up after remove/insert
Please see the logs below

Regards,
-kevin

......
Nov 25 21:47:25.015491 sonic INFO pmon#sfputil.py: PORT 31: unable to read PAGE0
Nov 25 21:47:25.015491 sonic INFO pmon#sfputil.py: PORT 31: Unable to initialize the module
Nov 25 21:48:06.595530 sonic INFO pmon#sfputil.py: PORT 31: unable to read PAGE0
Nov 25 21:48:06.595530 sonic INFO pmon#sfputil.py: PORT 31: Unable to initialize the module
Nov 25 21:49:36.731519 sonic INFO pmon#sfputil.py: PORT 31: unable to read PAGE0
Nov 25 21:49:36.731519 sonic INFO pmon#sfputil.py: PORT 31: Unable to initialize the module
Nov 25 21:50:11.383533 sonic INFO pmon#sfputil.py: PORT 31: unable to read PAGE0
Nov 25 21:50:11.383533 sonic INFO pmon#sfputil.py: PORT 31: Unable to initialize the module
Nov 25 21:50:39.103534 sonic INFO pmon#sfputil.py: PORT 31: unable to read PAGE0
Nov 25 21:50:39.103534 sonic INFO pmon#sfputil.py: PORT 31: Unable to initialize the module
Nov 25 21:51:34.543512 sonic INFO pmon#sfputil.py: PORT 31: unable to read PAGE0

File: device/celestica/x86_64-cel_silverstone-r0/plugins/sfputil.py
Comment (ds952811) [reply device/celestica/x86_64-cel_silverstone-r0/plugins/sfputil.py]: 1. We'll actually do nothing in the case of DAC, so I'm not sure why we need to keep the counter advanced? Do we want to have any other automatic recovery for DAC?

2. The reason why this is not handled in the insertion event is that, the I2C read could fail and it was observed during the QA test last week. However, we could add a inited flag to make sure this is only performed once and only once, let me update the code logic

Comment (zhenggen-xu) [reply device/celestica/x86_64-cel_silverstone-r0/plugins/sfputil.py]: 1- that is exactly my point, we should not advance the counter and then avoid the write here,  we should eliminate the counter advance in the code earlier so it won't even be a failure case.

Comment (ds952811) [reply device/celestica/x86_64-cel_silverstone-r0/plugins/sfputil.py]: copy that, and thanks, I'll update the code accordingly

File: device/celestica/x86_64-cel_silverstone-r0/plugins/sfputil.py
Comment (ds952811) [inline device/celestica/x86_64-cel_silverstone-r0/plugins/sfputil.py:459]: Ensure the module will be placed in LowPower mode

File: device/celestica/x86_64-cel_silverstone-r0/plugins/sfputil.py
Comment (ds952811) [inline device/celestica/x86_64-cel_silverstone-r0/plugins/sfputil.py:494]: Eoptolink CMIS4 took 6 seconds for config status to get populated to the register, and hence 10 sec should be sufficient

File: device/celestica/x86_64-cel_silverstone-r0/plugins/sfputil.py
Comment (ds952811) [inline device/celestica/x86_64-cel_silverstone-r0/plugins/sfputil.py:607]: Just for a record, this checker is necessary to prevent redundant CMIS4 init upon warm-reboot, otherwise significant packet drops will be observed.

File: platform/broadcom/sonic-platform-modules-cel/silverstone/modules/switchboard.c
Comment (ds952811) [inline platform/broadcom/sonic-platform-modules-cel/silverstone/modules/switchboard.c]: Revert the clock to high-speed, unless it's unlikely to be the root cause, and we actually have this enabled in ICOS, and I never got any issues for this

File: device/celestica/x86_64-cel_silverstone-r0/plugins/sfputil.py
Comment (zhenggen-xu) [inline device/celestica/x86_64-cel_silverstone-r0/plugins/sfputil.py]: This looks very similar to the code at line 656.  Can we make it as a function and reuse?

Comment (ds952811) [reply device/celestica/x86_64-cel_silverstone-r0/plugins/sfputil.py]: It's done, please check it out

**Comment by zhenggen-xu** [comment]
As you mentioned, please send the all clear flag once the QA is done.

**Comment by ds952811** [comment]
> As you mentioned, please send the all clear flag once the QA is done.

Although the hotfix in the last commit 'CLS Silverstone: sfputil.py: fix Eoptolink CMIS3 support' has already verified by the QA, we still need to generate a new build with this change, and I'm working on it, will let you know once it's ready.

**Comment by ds952811** [comment]
> > As you mentioned, please send the all clear flag once the QA is done.
> 
> Although the hotfix in the last commit 'CLS Silverstone: sfputil.py: fix Eoptolink CMIS3 support' has already verified by the QA, we still need to generate a new build with this change, and I'm working on it, will let you know once it's ready.

After the internal discussion with my manager, Babu, it sounds like you're not using the binary from Broadcom, and you will use only the image rebuilt from the source package.

In this case, you don't need to wait for the new build(image) from my end, please approve and merge the changes if no any concern

**Review by zhenggen-xu — APPROVED** [review]

**Pull request merged**
